### PR TITLE
[CUDA EP] RNN check if tf32 is allowed

### PR DESCRIPTION
**Pull request by gedoensmax**
Respect the use_tf32 flag.
@tianleiwu

### CODE REVIEW

File: onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.h [hunk 1]
```diff
@@ -40,10 +38,17 @@ class CudnnRNN {
...
              cudnnDropoutDescriptor_t cudnn_dropout_desc, cudnnDirectionMode_t cudnn_direction_model,
-             cudnnRNNMode_t rnn_mode, bool has_bias, cudnnDataType_t dataType) {
+             cudnnRNNMode_t rnn_mode, bool has_bias, cudnnDataType_t dataType, bool use_tf32) {
     if (!cudnn_rnn_desc_)
       CUDNN_RETURN_IF_ERROR(cudnnCreateRNNDescriptor(&cudnn_rnn_desc_));
 
+    cudnnMathType_t mathType = CUDNN_DEFAULT_MATH;
+    if (dataType == CUDNN_DATA_HALF) {
+      mathType = CUDNN_TENSOR_OP_MATH;
+    } else if (dataType == CUDNN_DATA_FLOAT && use_tf32) {
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.h:48]: CUDNN_DEFAULT_MATH allows TF32:
https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-graph-library.html#cudnnmathtype-t

What's extra conversion allowed for CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION?

May use the following logic to be consistent with other places:
```
else if (args.params.data_type == CUDNN_DATA_FLOAT && !use_tf32) {
    perf_results[0].mathType = CUDNN_FMA_MATH;
}

Comment (gedoensmax) [reply onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.h:48]: Given the docs [here](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-adv-library.html#cudnnsetrnndescriptor-v8) `CUDNN_FMA_MATH` would not be an option for RNN's. Allow conversion would enable storing intermediate steps in fp16 which has the same mantissa as tf32.

Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.h:48]: I saw the [doc](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-adv-library.html#cudnnsetrnndescriptor-v8) mentioned that 

When dataType is CUDNN_DATA_FLOAT, the mathType parameter can be CUDNN_DEFAULT_MATH or CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION. When the latter settings are used, original weights and intermediate results will be **down-converted to CUDNN_DATA_HALF** before they are used in another recursive iteration.

Comment (gedoensmax) [reply onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.h:48]: Turns out you were right `CUDNN_FMA_MATH` is in fact available on RNN although it is not mentioned in the docs explicitly.

**Review by tianleiwu — DISMISSED** [review]

**Comment by tianleiwu** [comment]
Need run `lintrunner -a` to fix python format.

**Review by tianleiwu — APPROVED** [review]

**Comment by gedoensmax** [comment]
@tianleiwu Should I rebase to master or will the be merged in the current state ?

**Comment by tianleiwu** [comment]
@gedoensmax, Let me try re-run the MacOS CI Pipeline. If it passes, we can merge it without rebasing.

**Pull request merged**